system_zero/src/permutation_unit.rs

```diff
@@ -72,15 +195,152 @@ impl<F: RichField + Extendable<D>, const D: usize> SystemZero<F, D> {
+
+            state = F::mds_layer_recursive(builder, &state);
+        }
+
+        for r in 0..HALF_N_FULL_ROUNDS {
+            F::constant_layer_recursive(
+                builder,
+                &mut state,
+                HALF_N_FULL_ROUNDS + N_PARTIAL_ROUNDS + r,
+            );
+
+            for i in 0..SPONGE_WIDTH {
+                let state_cubed = builder.cube_extension(state[i]);
+                let diff = builder.sub_extension(
+                    state_cubed,
+                    local_values[col_permutation_full_second_mid_sbox(r, i)],
+                );
+                yield_constr.one(builder, diff);
+                let state_cubed = local_values[col_permutation_full_second_mid_sbox(r, i)];
+                state[i] = builder.mul_many_extension(&[state[i], state_cubed, state_cubed]);
+                // Form state ** 7.
+            }
+
+            state = F::mds_layer_recursive(builder, &state);
+
+            for i in 0..SPONGE_WIDTH {
+                let diff = builder.sub_extension(
+                    state[i],
+                    local_values[col_permutation_full_second_after_mds(r, i)],
+                );
+                yield_constr.one(builder, diff);
+                state[i] = local_values[col_permutation_full_second_after_mds(r, i)];
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use plonky2::field::field_types::{Field, PrimeField};
+    use plonky2::field::goldilocks_field::GoldilocksField;
+    use plonky2::hash::poseidon::Poseidon;
+    use starky::constraint_consumer::ConstraintConsumer;
+    use starky::vars::StarkEvaluationVars;
+
+    use crate::column_layout::{col_permutation_input, col_permutation_output, NUM_COLUMNS};
+    use crate::permutation_unit::SPONGE_WIDTH;
+    use crate::public_input_layout::NUM_PUBLIC_INPUTS;
+    use crate::system_zero::SystemZero;
+
+    #[test]
+    fn generate_eval_consistency() {
+        const D: usize = 1;
+        type F = GoldilocksField;
+
+        let mut values = [F::default(); NUM_COLUMNS];
+        SystemZero::<F, D>::generate_permutation_unit(&mut values);
+
+        let vars = StarkEvaluationVars {
+            local_values: &values,
+            next_values: &[F::default(); NUM_COLUMNS],
+            public_inputs: &[F::default(); NUM_PUBLIC_INPUTS],
+        };
+
+        let mut constrant_consumer = ConstraintConsumer::new(
+            vec![GoldilocksField(2), GoldilocksField(3), GoldilocksField(5)],
+            GoldilocksField::ONE,
+            GoldilocksField::ONE,
+        );
+        SystemZero::<F, D>::eval_permutation_unit(vars, &mut constrant_consumer);
+        for &acc in &constrant_consumer.constraint_accs {
+            assert_eq!(acc, GoldilocksField::ZERO);
         }
     }
+
+    #[test]
+    fn poseidon_result() {
+        const D: usize = 1;
+        type F = GoldilocksField;
+
+        let mut state = [F::default(); SPONGE_WIDTH];
```

Comment:
Maybe use `F::rand_arr::<NUM_COLUMNS>()` to make sure it passes with other inputs?

Reply:
Good idea!